Bug with --testdox option at v7.4.0

| Q | A
| --- | ---
| ParaTest version | 7.4.0
| PHPUnit version | 10.5.10
| PHP version | 8.2.11

### Summary

Hi, good evening,

There's a bug with the new v7.4.0, if I downgrade to v7.3.1 it works as expected.
It seems the bug is related to the option `--testdox` at v7.4.0, if I remove that parameter works as expected.

Feel free to ask if you want me to test anything. 😄

Thanks!

### Current behavior

#### ParaTest v7.4.0 with --testdox (Laravel Artisan)

```
/var/www/html # php artisan test --without-tty --parallel --testdox
ParaTest v7.4.0 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     4
Runtime:       PHP 8.2.11
Configuration: /var/www/html/phpunit.xml


In WorkerCrashedException.php line 41:

  The test "PARATEST='1' TEST_TOKEN='1' UNIQUE_TEST_TOKEN='1_65c10b9d06124' t
  ests/Unit/ExampleTest.php" failed.

  Exit Code: 255(Unknown error)

  Working directory: /var/www/html

  Output:
  ================


  Error Output:
  ================
  PHP Warning:  Undefined array key "phpunit-argv" in /var/www/html/vendor/br
  ianium/paratest/bin/phpunit-wrapper.php on line 47
  PHP Fatal error:  Uncaught TypeError: unserialize(): Argument #1 ($data) mu
  st be of type string, null given in /var/www/html/vendor/brianium/paratest/
  bin/phpunit-wrapper.php:47
  Stack trace:
  #0 /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php(47): unse
  rialize(NULL, Array)
  #1 /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php(79): {clo
  sure}()
  #2 {main}
    thrown in /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php
  on line 47


paratest [--functional] [-m|--max-batch-size MAX-BATCH-SIZE] [--no-test-tokens] [--passthru-php PASSTHRU-PHP] [-p|--processes PROCESSES] [--runner RUNNER] [--tmp-dir TMP-DIR] [-v|--verbose] [--bootstrap BOOTSTRAP] [-c|--configuration CONFIGURATION] [--no-configuration] [--cache-directory CACHE-DIRECTORY] [--testsuite TESTSUITE] [--exclude-testsuite EXCLUDE-TESTSUITE] [--group GROUP] [--exclude-group EXCLUDE-GROUP] [--filter FILTER] [--process-isolation] [--strict-coverage] [--strict-global-state] [--disallow-test-output] [--dont-report-useless-tests] [--stop-on-defect] [--stop-on-error] [--stop-on-failure] [--stop-on-warning] [--stop-on-risky] [--stop-on-skipped] [--stop-on-incomplete] [--fail-on-incomplete] [--fail-on-risky] [--fail-on-skipped] [--fail-on-warning] [--order-by ORDER-BY] [--random-order-seed RANDOM-ORDER-SEED] [--colors [COLORS]] [--no-progress] [--display-incomplete] [--display-skipped] [--display-deprecations] [--display-errors] [--display-notices] [--display-warnings] [--teamcity] [--testdox] [--log-junit LOG-JUNIT] [--log-teamcity LOG-TEAMCITY] [--coverage-clover COVERAGE-CLOVER] [--coverage-cobertura COVERAGE-COBERTURA] [--coverage-crap4j COVERAGE-CRAP4J] [--coverage-html COVERAGE-HTML] [--coverage-php COVERAGE-PHP] [--coverage-text [COVERAGE-TEXT]] [--coverage-xml COVERAGE-XML] [--coverage-filter COVERAGE-FILTER] [--no-coverage] [--] [<path>]
```
------

#### ParaTest v7.4.0 with --testdox (Paratest binary)

```
/var/www/html # vendor/bin/paratest --testdox
ParaTest v7.4.0 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     4
Runtime:       PHP 8.2.11
Configuration: /var/www/html/phpunit.xml


In WorkerCrashedException.php line 41:

  The test "PARATEST='1' TEST_TOKEN='1' UNIQUE_TEST_TOKEN='1_65c10cea49f7e' tests/Unit/ExampleTest.php" failed.

  Exit Code: 255(Unknown error)

  Working directory: /var/www/html

  Output:
  ================


  Error Output:
  ================
  PHP Warning:  Undefined array key "phpunit-argv" in /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php on line 47
  PHP Fatal error:  Uncaught TypeError: unserialize(): Argument #1 ($data) must be of type string, null given in /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php:47
  Stack trace:
  #0 /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php(47): unserialize(NULL, Array)
  #1 /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php(79): {closure}()
  #2 {main}
    thrown in /var/www/html/vendor/brianium/paratest/bin/phpunit-wrapper.php on line 47
```

### How to reproduce: command, code and error stack trace

I'm using the following command:

```bash
php artisan test --without-tty --parallel --testdox
```

### Expected behavior

#### ParaTest v7.3.1 with --testdox (Laravel Artisan)

```
/var/www/html # php artisan test --without-tty --parallel --testdox
ParaTest v7.3.1 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     4
Runtime:       PHP 8.2.11
Configuration: /var/www/html/phpunit.xml

........................                                          24 / 24 (100%)

Time: 00:02.250, Memory: 26.00 MB

Example (Tests\Unit\Example)
 ✔ That true is true

Locale (Tests\Feature\Locale)
 ✔ The application returns a successful response
 ✔ The application redirects to default path
 ✔ The application returns a successful response in french
 ✔ The application returns a successful response in english

Authentication (Tests\Feature\Auth\Authentication)
 ✔ Login screen can be rendered
 ✔ Users can authenticate using the login screen
 ✔ Users can logout
 ✔ Users can not authenticate with invalid password
 ✔ Users can not authenticate with inactive account
 ✔ Users gets redirected to business
 ✔ User gets redirected to preferred locale after login

Password Reset (Tests\Feature\Auth\PasswordReset)
 ✔ Reset password link screen can be rendered
 ✔ Reset password link can be requested
 ✔ Reset password screen can be rendered
 ✔ Password can be reset with valid token

Create User (Tests\Feature\Commands\CreateUser)
 ✔ Can create user
 ✔ Cannot create user with existing email
 ✔ Cannot create user with existing username
 ✔ Can create user with password
 ✔ Can not create user with password mismatch
 ✔ Can login with created user
 ✔ Can create user with notification
 ✔ Can create user without notification

OK (24 tests, 106 assertions)
```

------

#### ParaTest v7.4.0 without --testdox  (Laravel Artisan)

```
/var/www/html # php artisan test --without-tty --parallel
ParaTest v7.4.0 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     4
Runtime:       PHP 8.2.11
Configuration: /var/www/html/phpunit.xml

........................                                          24 / 24 (100%)

Time: 00:02.152, Memory: 26.00 MB

OK (24 tests, 106 assertions)
```

Sure!
Here you are https://github.com/sertxudev/paratest-bug

It's a Laravel fresh installation, I've only changed the database from MySQL to SQLite.
Running the default tests fails.

#### vendor\bin\paratest --testdox

```
Z:\Website_Developments\home\laravel\data>vendor\bin\paratest --testdox
ParaTest v7.4.0 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     6
Runtime:       PHP 8.2.3
Configuration: Z:\Website_Developments\home\laravel\data\phpunit.xml


In WorkerCrashedException.php line 41:

  The test "PARATEST="1" TEST_TOKEN="1" UNIQUE_TEST_TOKEN="1_65c212950b915" tests\Unit\ExampleTest.php" failed.

  Exit Code: 255(Unknown error)

  Working directory: Z:\Website_Developments\home\laravel\data

  Output:
  ================

  Fatal error: Uncaught AssertionError: assert(isset($getopt['phpunit-argv']) && is_string($getopt['phpunit-argv']))
  in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php:46
  Stack trace:
  #0 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(46): assert(false, 'a
  ssert(isset($g...')
  #1 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(79): {closure}()
  #2 {main}
    thrown in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php on line 46


  Error Output:
  ================
  PHP Fatal error:  Uncaught AssertionError: assert(isset($getopt['phpunit-argv']) && is_string($getopt['phpunit-argv
  '])) in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php:46
  Stack trace:
  #0 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(46): assert(false, 'a
  ssert(isset($g...')
  #1 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(79): {closure}()
  #2 {main}
    thrown in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php on line 46
```

#### php artisan test --parallel --testdox

```
Z:\Website_Developments\home\laravel\data>php artisan test --parallel --testdox
ParaTest v7.4.0 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     6
Runtime:       PHP 8.2.3
Configuration: Z:\Website_Developments\home\laravel\data\phpunit.xml


In WorkerCrashedException.php line 41:

  The test "PARATEST="1" TEST_TOKEN="1" UNIQUE_TEST_TOKEN="1_65c2130ac243c" tests\Unit\ExampleTest.php" failed.

  Exit Code: 255(Unknown error)

  Working directory: Z:\Website_Developments\home\laravel\data

  Output:
  ================

  Fatal error: Uncaught AssertionError: assert(isset($getopt['phpunit-argv']) && is_string($getopt['phpunit-argv'])) in Z:\Website_Developments\home\laravel\d
  ata\vendor\brianium\paratest\bin\phpunit-wrapper.php:46
  Stack trace:
  #0 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(46): assert(false, 'assert(isset($g...')
  #1 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(79): {closure}()
  #2 {main}
    thrown in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php on line 46


  Error Output:
  ================
  PHP Fatal error:  Uncaught AssertionError: assert(isset($getopt['phpunit-argv']) && is_string($getopt['phpunit-argv'])) in Z:\Website_Developments\home\lara
  vel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php:46
  Stack trace:
  #0 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(46): assert(false, 'assert(isset($g...')
  #1 Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php(79): {closure}()
  #2 {main}
    thrown in Z:\Website_Developments\home\laravel\data\vendor\brianium\paratest\bin\phpunit-wrapper.php on line 46
```

> [!NOTE]
> Downgrading to v7.3.1 works with both commands

#### vendor\bin\paratest --testdox (v7.3.1)

```
Z:\Website_Developments\home\laravel\data>vendor\bin\paratest --testdox
ParaTest v7.3.1 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     6
Runtime:       PHP 8.2.3
Configuration: Z:\Website_Developments\home\laravel\data\phpunit.xml

..                                                                 2 / 2 (100%)

Time: 00:00.598, Memory: 10.00 MB

Example (Tests\Unit\Example)
 ✔ That true is true

Example (Tests\Feature\Example)
 ✔ The application returns a successful response

OK (2 tests, 2 assertions)
```

#### php artisan test --parallel --testdox (v7.3.1)

```
Z:\Website_Developments\home\laravel\data>php artisan test --parallel --testdox
ParaTest v7.3.1 upon PHPUnit 10.5.10 by Sebastian Bergmann and contributors.

Processes:     6
Runtime:       PHP 8.2.3
Configuration: Z:\Website_Developments\home\laravel\data\phpunit.xml

..                                                                 2 / 2 (100%)

Time: 00:00.766, Memory: 22.00 MB

Example (Tests\Unit\Example)
 ✔ That true is true

Example (Tests\Feature\Example)
 ✔ The application returns a successful response

OK (2 tests, 2 assertions)
```

--- a/bin/phpunit-wrapper.php
+++ b/bin/phpunit-wrapper.php
@@ -14,6 +14,7 @@ use ParaTest\WrapperRunner\WrapperWorker;
         'teamcity-file:',
         'testdox-file:',
         'testdox-color',
+        'testdox-columns:',
         'phpunit-argv:',
     ]);
 
@@ -55,7 +56,7 @@ use ParaTest\WrapperRunner\WrapperWorker;
         $getopt['teamcity-file'] ?? null,
         $getopt['testdox-file'] ?? null,
         isset($getopt['testdox-color']),
-        $getopt['testdox-columns'] ?? null,
+        isset($getopt['testdox-columns']) ? (int) $getopt['testdox-columns'] : null,
     );
 
     while (true) {

--- a/bin/phpunit-wrapper.php
+++ b/bin/phpunit-wrapper.php
@@ -42,7 +42,7 @@ use ParaTest\WrapperRunner\WrapperWorker;
     assert(isset($getopt['testresult-file']) && is_string($getopt['testresult-file']));
     assert(!isset($getopt['teamcity-file']) || is_string($getopt['teamcity-file']));
     assert(!isset($getopt['testdox-file']) || is_string($getopt['testdox-file']));
-    assert(!isset($getopt['testdox-columns']) || is_int($getopt['testdox-columns']));
+    assert(!isset($getopt['testdox-columns']) || $getopt['testdox-columns'] === (string) (int) $getopt['testdox-columns']);
 
     assert(isset($getopt['phpunit-argv']) && is_string($getopt['phpunit-argv']));
     $phpunitArgv = unserialize($getopt['phpunit-argv'], ['allowed_classes' => false]);

--- a/composer.json
+++ b/composer.json
@@ -43,7 +43,7 @@
         "phpunit/php-code-coverage": "^10.1.11 || ^11.0.0",
         "phpunit/php-file-iterator": "^4.1.0 || ^5.0.0",
         "phpunit/php-timer": "^6.0.0 || ^7.0.0",
-        "phpunit/phpunit": "^10.5.9 || ^11.0.0",
+        "phpunit/phpunit": "^10.5.9 || ^11.0.2",
         "sebastian/environment": "^6.0.1 || ^7.0.0",
         "symfony/console": "^6.4.3 || ^7.0.3",
         "symfony/process": "^6.4.3 || ^7.0.3"

--- a/test/TestBase.php
+++ b/test/TestBase.php
@@ -59,6 +59,10 @@ abstract class TestBase extends TestCase
             $argv['--tmp-dir'] = $this->tmpDir;
         }
 
+        if (! isset($argv['--passthru-php'])) {
+            $argv['--passthru-php'] = "'-d' 'zend.assertions=1'";
+        }
+
         $input = new ArrayInput($argv, $inputDefinition);
 
         return Options::fromConsoleInput($input, $cwd ?? __DIR__);

--- a/test/Unit/OptionsTest.php
+++ b/test/Unit/OptionsTest.php
@@ -71,7 +71,7 @@ final class OptionsTest extends TestBase
         self::assertNotEmpty($options->phpunitOptions);
         self::assertSame(0, $options->maxBatchSize);
         self::assertFalse($options->noTestTokens);
-        self::assertEmpty($options->passthruPhp);
+        self::assertSame(['-d', 'zend.assertions=1'], $options->passthruPhp);
         self::assertStringContainsString('phpunit', $options->phpunit);
         self::assertSame(PROCESSES_FOR_TESTS, $options->processes);
         self::assertSame('WrapperRunner', $options->runner);

--- a/test/Unit/WrapperRunner/WrapperRunnerTest.php
+++ b/test/Unit/WrapperRunner/WrapperRunnerTest.php
@@ -263,6 +263,27 @@ final class WrapperRunnerTest extends TestBase
         );
     }
 
+    public function testTestdoxOutput(): void
+    {
+        $this->bareOptions['path']      = $this->fixture('common_results');
+        $this->bareOptions['--testdox'] = true;
+
+        $result = $this->runRunner();
+
+        $format = file_get_contents(__DIR__ . '/fixtures/common_results_testdox_output');
+        self::assertNotFalse($format);
+
+        $output = $result->output;
+        $output = preg_replace("/^Processes:     \\d+\nRuntime:       PHP \\d+.\\d+.\\d+(-\w+)?\n\n/", '', $output, 1, $count);
+        self::assertSame(1, $count);
+        self::assertNotNull($output);
+
+        self::assertStringMatchesFormat(
+            self::sorted($format),
+            self::sorted($output),
+        );
+    }
+
     public function testExitCodesPathWithoutTests(): void
     {
         $this->bareOptions['path'] = $this->fixture('no_tests');

--- a/test/Unit/WrapperRunner/fixtures/common_results_testdox_output
+++ b/test/Unit/WrapperRunner/fixtures/common_results_testdox_output
@@ -1,0 +1,42 @@
+
+EFIRS.W                                                             7 / 7 (100%)
+
+Time: %s, Memory: %s MB
+
+Error (ParaTest\Tests\fixtures\common_results\Error)
+ ✘ Error
+   │
+   │ RuntimeException: Error here!
+   │
+   │ %s/test/fixtures/common_results/ErrorTest.php:15
+   │
+
+Incomplete (ParaTest\Tests\fixtures\common_results\Incomplete)
+ ∅ Incomplete
+   │
+   │
+   │ %s/test/fixtures/common_results/IncompleteTest.php:14
+   │
+
+Skipped (ParaTest\Tests\fixtures\common_results\Skipped)
+ ↩ Skipped
+
+Warning (ParaTest\Tests\fixtures\common_results\Warning)
+ ⚠ Warning
+
+Failure (ParaTest\Tests\fixtures\common_results\Failure)
+ ✘ Failure
+   │
+   │ Failed asserting that false is true.
+   │
+   │ %s/test/fixtures/common_results/FailureTest.php:14
+   │
+
+Risky (ParaTest\Tests\fixtures\common_results\Risky)
+ ⚠ Risky
+
+Success (ParaTest\Tests\fixtures\common_results\Success)
+ ✔ Success
+
+ERRORS!
+Tests: %s

